src/compiler/Restler.Compiler/Grammar.fs

```diff
@@ -13,23 +13,31 @@ module Tree =
         | LeafNode of 'LeafData
         | InternalNode of 'INodeData * Tree<'LeafData,'INodeData> seq
 
-    let rec cata fLeaf fNode (tree:Tree<'LeafData,'INodeData>) :'r =
-        let recurse = cata fLeaf fNode
-        match tree with
-        | LeafNode leafInfo ->
-            fLeaf leafInfo
-        | InternalNode (nodeInfo,subtrees) ->
-            fNode nodeInfo (subtrees |> Seq.map recurse)
 
     /// Same as cata, but allows passing a parent context
-    let rec cataCtx fLeaf fNode fCtx ctx (tree:Tree<'LeafData,'INodeData>) :'r =
-        let recurse = cataCtx fLeaf fNode fCtx
-        match tree with
-        | LeafNode leafInfo ->
-            fLeaf ctx leafInfo
-        | InternalNode (nodeInfo,subtrees) ->
-            let newCtx = fCtx ctx nodeInfo
-            fNode ctx nodeInfo (subtrees |> Seq.map (recurse newCtx))
+    let cataCtx fLeaf fNode fCtx ctx (tree:Tree<'LeafData,'INodeData>) :'r =
+        let rec contMap newCtx (trees: seq<Tree<'LeafData,'INodeData>>) (conts: seq<'r> -> 'r): 'r =
+            if Seq.isEmpty trees then
+                conts []
+            else
+                let h = Seq.head trees
+                contMap newCtx (Seq.tail trees) (fun ts ->
+                    r newCtx h (fun s -> conts (Seq.append [s] ts))
+                )
+        and r ctx tree cont =
```

Comment:
>r [](start = 12, length = 1)

could you please give this a more descriptive name?  Or, at least a different name for readability (since 'r is already used above)?